test/e2e/apps/job.go

```diff
@@ -344,6 +344,57 @@ var _ = SIGDescribe("Job", func() {
 		}
 	})
 
+	ginkgo.It("should recreate pods only after they have failed if pod replacement policy is set to Failed", func(ctx context.Context) {
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJob("notTerminate", "pod-recreate-failed", v1.RestartPolicyNever, 1, 1, nil, 1)
+		job.Spec.Template.Finalizers = []string{"fake.example.com/blockDeletion"}
+		job.Spec.PodReplacementPolicy = ptr.To(batchv1.Failed)
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+		defer func() {
+			framework.ExpectNoError(removePodsFinalizer(ctx, f.ClientSet, f.Namespace.Name), "failed to remove finalizer from pods in namespace: %s", f.Namespace.Name)
+		}()
+
+		err = e2ejob.WaitForJobPodsRunning(ctx, f.ClientSet, f.Namespace.Name, job.Name, 1)
+		framework.ExpectNoError(err, "failed to wait for job pod to become running in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Deleting job pod")
+		pods, err := e2ejob.GetJobPods(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to get pod list for job %s in namespace: %s", job.Name, f.Namespace.Name)
+
+		framework.ExpectNoError(e2epod.DeletePodsWithGracePeriod(ctx, f.ClientSet, pods.Items, 0), "failed to delete pods in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Ensuring pod does not get recreated while it is in terminating state")
+		err = e2ejob.WaitForJobState(ctx, f.ClientSet, f.Namespace.Name, job.Name, f.Timeouts.PodDelete, func(job *batchv1.Job) string {
+			if job.Status.Active == 0 && job.Status.Failed == 0 && *job.Status.Terminating == 1 {
+				return ""
```

Comment:
what prevents the pod from finishing too fast, so that it's in a terminating state before reaching Failed?

Reply:
hmmm good question, kubelet needs some time to transition it, but maybe there is a small chance it happens super fast and test fails. I don't have a good answer, do you have ideas?

Reply:
That's why we had the grace period in the pod itself... the tricky part is forcing a delete after sending a SIGTERM.